### PR TITLE
Implement InstanceChangeHandler in former TaskUpdateStepImpls

### PR DESCRIPTION
**Pull request by meichstedt**
added implementation for new `InstanceChange` to the steps that currently propagate `TaskChanged`. Will remove old stuff and rename steps in a follow up PR

---
This change is [Reviewable](https://reviewable.io/reviews/mesosphere/marathon/4342)

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala [hunk 1]
```diff
@@ -22,18 +22,21 @@ trait InstanceChangeHandler {
   * An event notifying of an [[Instance]] change.
   */
 sealed trait InstanceChange {
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala]: extends Product with Serializable: (aka make all case classes/case objects underneath)

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala [hunk 2]
```diff
@@ -125,6 +120,17 @@ private[impl] class LaunchQueueActor(
...
 
+  private[this] def receiveInstanceUpdateToSuspendedActor: Receive = {
+    case update: InstanceChange if suspendedLauncherPathIds(update.runSpecId) =>
+      // Do not defer. If an AppTaskLauncherActor restarts, it retrieves a new task list.
+      // If we defer this, there is a potential deadlock (resolved by timeout):
+      //   * AppTaskLauncher waits for in-flight tasks
+      //   * TaskOp gets processed and one of the update steps calls this here
+      //   * ... blocked until timeout ...
+      //   * The task launch notification (that the AppTaskLauncherActor waits for) gets sent to the actor
+      sender() ! None
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: None?

Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: Maybe better as Option.empty[T] for readability sake...

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala [hunk 3]
```diff
@@ -17,12 +19,37 @@ import scala.concurrent.Future
...
+      case _: InstanceStatus.Terminal =>
+        val runSpecId = update.runSpecId
+        val instanceId = update.id
+        val state = update.status
+        log.info(s"initiating a scale check for runSpec [$runSpecId] due to [$instanceId] $state")
+        // TODO(PODS):
+        //    1) SchedulerActor is awful
+        //    2) yet it handles runSpecs locked due to deployments
+        //    3) if we still use it, we should rename the Message and make the SchedulerActor generic
+        schedulerActor ! ScaleApp(runSpecId)
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala]: Doesn't ScaleApp work for both? Or am I just thinking of plans.

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala (hunk 2 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: IIRC previously the message was stashed and eventually the ask timed out … basically what the comment is saying. What bothers me a little is that the QueuedInstanceInfo, which is sent back in the usual case, is not used by anyone, only in test classes (which could as well do a `Get`in order to retrieve the count). I'd be happy to change affected code to not ask or simply expect a Done or whatever – we're mostly mapping the response to Done or Unit.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/TaskStatusEmitterPublishStepImpl.scala [hunk 4]
```diff
@@ -12,13 +14,22 @@ import scala.concurrent.Future
 /**
   * Forward the update to the taskStatusEmitter.
   */
+// TODO(PODS): is the emitter stuff still in use?
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/TaskStatusEmitterPublishStepImpl.scala]: I'm hoping we can remove rxScala and use a lot more of akka streaming. I think a bunch of stuff we're currently either using even stream for or actors could be replaced with streaming.

**Comment by jasongilanfarr** [comment]
Mostly minor comments. 👍 with all the "Done" usage. Would love more async/await (or Source[])

File: src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala (hunk 1 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala]: sure, will do

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala (hunk 2 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: Oh- I should have been more explicit- it just wasn't clear in this particular call site what the inner type of the option is.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala (hunk 3 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala]: I think it works for both, and if not, the easiest thing would be to make it work for both

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala (hunk 2 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala]: Ah – I'd be ok with `Option.empty[QueuedInstanceInfo]` as well

**Comment by jdef** [comment]
Review status: 0 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala, line 31 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KREaFf_MzzDBKrUlKFJ:-KREaFf_MzzDBKrUlKFK:bvrkew5) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala#L31)):*
> ```Scala
>   /** Status of the [[Instance]] */
>   // TODO(PODS): We might want to transport health information in the status
>   // TODO(PODS): who carries the version? instance or InstanceStatus?
> ```

Pod has an immutable version. So should each instance status (it's possible that some instance's are running with older versions of some pod-instance-spec)

---

*[src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala, line 242 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KREdie8ulMZlMoV4Teu:-KREdie8ulMZlMoV4Tev:btullhi) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala#L242)):*
> ```Scala
>       op match {
>         // only increment for launch ops, not for reservations:
>         case _: InstanceOp.LaunchTask => instancesToLaunch += 1
> ```

TODO: for pods InstanceOp.LaunchTaskGroup?

---

*[src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala, line 403 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KREdv5GkSu7S9688p4w:-KREdv5GkSu7S9688p4x:btullhi) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala#L403)):*
> ```Scala
>       instanceOp match {
>         // only decrement for launched instances, not for reservations:
>         case _: InstanceOp.LaunchTask => instancesToLaunch -= 1
> ```

TODO: for pods InstanceOp.LaunchTaskGroup?

---

*[src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala, line 465 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KREegEvzV-tFRK8jMUl:-KREegEvzV-tFRK8jMUm:b-jfmn44) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala#L465)):*
> ```Scala
>     private[this] val myselfAsOfferMatcher: OfferMatcher = {
>       //set the precedence only, if this app is resident
>       // TODO (pods): how shall we handle residency?
> ```

we don't handle residency for the pods MVP

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mesosphere/marathon/4342)*

**Comment by jdef** [comment]
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mesosphere/marathon/4342)*

**Comment by meichstedt** [comment]
Review status: 16 of 17 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala, line 24 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KRE_gW-i-xRJaUUdB7n-r1-24:-KREjTfeoOhPD9Q6JXMP:b-896fix) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala#L24)):*
<details><summary><i>Previously, meichstedt (Matthias Eichstedt) wrote…</i></summary>
> sure, will do
</details>
Done.

---

*[src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala, line 31 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KREaFf_MzzDBKrUlKFJ:-KREmj6HL2B_zECA6umz:bw4kda6) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/instance/update/InstanceChangeHandler.scala#L31)):*
<details><summary><i>Previously, jdef (James DeFelice) wrote…</i></summary>
> Pod has an immutable version. So should each instance status (it's possible that some instance's are running with older versions of some pod-instance-spec)
</details>
Ack; I added the runSpecVersion to `InstanceState`

---

*[src/main/scala/mesosphere/marathon/core/task/update/impl/steps/TaskStatusEmitterPublishStepImpl.scala, line 17 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4342#-KRE_gW02TeMXREyu6tL-r1-17:-KREo0ZSmp6l7lfalFct:btelcpo) ([raw file](https://github.com/mesosphere/marathon/blob/bf2abd9b73bdaec7ad7cd7de63f4ddf624cd6cd3/src/main/scala/mesosphere/marathon/core/task/update/impl/steps/TaskStatusEmitterPublishStepImpl.scala#L17)):*
<details><summary><i>Previously, jasongilanfarr (Jason Gilanfarr) wrote…</i></summary>
> I'm hoping we can remove rxScala and use a lot more of akka streaming. I think a bunch of stuff we're currently either using even stream for or actors could be replaced with streaming.
</details>
I'd like to get rid of this specific piece on master; the affected classes can use the eventStream like we do elsewhere; otherwise I'd need to adjust this here to be able to send either a mocked TaskChanged or widen the type to send InstanceChange or whatever. It's only used to increase LaunchTokens afaict.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mesosphere/marathon/4342)*

**Pull request merged**